File: frontend/src/lib/types/auditLog.ts
/**
 * Tipos do dominio de Audit Log usados pelo frontend (Wave 6, Componente 18).
 * Espelho fiel dos schemas Pydantic em backend/app/domain/schemas/audit_log.py.
 *
 * Acesso restrito ao perfil 3Studio (is_admin=true) — defesa em profundidade
 * via middleware FastAPI + RLS pol_audit_select + guard de menu.
 */

import type { Rota, Setor, StatusProva } from "./prova";

// ─── Acoes conhecidas (universo aberto) ──────────────────────────────────

/**
 * Lista das acoes conhecidas hoje. O backend aceita qualquer string <= 100
 * chars — migrations futuras podem adicionar valores. A UI exibe o valor
 * cru se nao estiver no mapa de labels.
 */
export const ACOES_CONHECIDAS = [
  "criar_prova",
  "escanear_prova",
  "transitar_status",
  "reiniciar_ciclo",
  "atualizar_configuracao",
  "REPORT_EXPORTED",
] as const;

export type AcaoConhecida = (typeof ACOES_CONHECIDAS)[number];

/** Tipo aberto: aceita acoes futuras nao listadas. */
export type Acao = AcaoConhecida | string;

/** Labels pt-BR para acoes conhecidas. UI faz fallback para o valor cru. */
export const ACAO_LABELS: Record<AcaoConhecida, string> = {
  criar_prova: "Criar prova",
  escanear_prova: "Escanear QR Code",
  transitar_status: "Mudar status",
  reiniciar_ciclo: "Reiniciar ciclo",
  atualizar_configuracao: "Atualizar configuracao",
  REPORT_EXPORTED: "Exportar relatorio",
};

/** Helper para formatar a acao para exibicao. */
export function formatAcao(acao: string): string {
  if (acao in ACAO_LABELS) {
    return ACAO_LABELS[acao as AcaoConhecida];
  }
  return acao;
}

// ─── Itens da listagem e detalhe ─────────────────────────────────────────

/** Linha individual da listagem de audit_logs.
 *
 * `assinatura_digital` (BYTEA) NUNCA e exposta — backend impede.
 */
export interface AuditLogItemResponse {
  id: string;
  acao: string;
  prova_id: string | null;
  prova_nro_requerimento: string | null;
  usuario_id: string;
  usuario_nome: string;
  usuario_setor: Setor;
  detalhes_json: Record<string, unknown> | null;
  ip_address: string | null;
  user_agent: string | null;
  created_at: string;
}

/** Resumo de movimentacao relacionada a um audit_log de transicao.
 *
 * Aparece no detalhe quando acao for transitar_status ou reiniciar_ciclo
 * E o backend conseguir matchear via prova_id + status_novo + ciclo +
 * janela ±5s.
 *
 * `assinatura_digital_presente` e o unico vestigio do BYTEA — confirma
 * que houve assinatura conforme RN-003 sem expor o conteudo.
 */
export interface MovimentacaoSnapshot {
  id: string;
  status_anterior: StatusProva;
  status_novo: StatusProva;
  motivo_reprovacao: string | null;
  ciclo: number;
  rota_no_momento: Rota | null;
  assinatura_digital_presente: boolean;
  created_at: string;
}

/** Detalhe individual com enriquecimento opcional. */
export interface AuditLogDetailResponse extends AuditLogItemResponse {
  movimentacao_relacionada: MovimentacaoSnapshot | null;
}

/** Response paginado da listagem (e do by-prova, que tambem usa esse shape). */
export interface AuditLogListResponse {
  items: AuditLogItemResponse[];
  total: number;
  page: number;
  page_size: number;
}

// ─── Tipo de evento semantico (UX A2) ────────────────────────────────────

/** Categorias semanticas de alto nivel para filtro `tipo_evento`.
 *
 * Mapeia para combinacoes de `acao` + `detalhes_json` no backend.
 * 'todos' (e null) = sem filtro.
 */
export type TipoEvento =
  | "todos"
  | "reprovacao"
  | "reinicio"
  | "cancelamento"
  | "criacao"
  | "admin";

export const TIPO_EVENTO_OPTIONS: readonly TipoEvento[] = [
  "todos",
  "reprovacao",
  "reinicio",
  "cancelamento",
  "criacao",
  "admin",
] as const;

export const TIPO_EVENTO_LABELS: Record<TipoEvento, string> = {
  todos: "Todos os eventos",
  reprovacao: "Apenas reprovacoes",
  reinicio: "Apenas reinicios de ciclo",
  cancelamento: "Apenas cancelamentos",
  criacao: "Apenas criacoes de prova",
  admin: "Mudancas administrativas",
};

// ─── Ordenacao (UX B4) ───────────────────────────────────────────────────

/** Colunas pelas quais a listagem pode ser ordenada (whitelist do backend). */
export type OrderBy = "created_at" | "acao" | "usuario_nome";

export const ORDER_BY_LABELS: Record<OrderBy, string> = {
  created_at: "Data e hora",
  acao: "Acao",
  usuario_nome: "Ator",
};

// ─── Page size (UX B2) ───────────────────────────────────────────────────

export const PAGE_SIZE_OPTIONS: readonly number[] = [25, 50, 100, 200] as const;

// ─── Filtros (query params) ──────────────────────────────────────────────

/** Filtros aceitos pelo GET /api/v1/audit-log.
 *
 * Espelha AuditLogListQuery do backend. Todos opcionais — se omitidos,
 * o backend usa defaults (page=1, page_size=50, sort=desc,
 * order_by=created_at, sem filtros).
 */
export interface AuditLogFilters {
  page: number;
  page_size: number;
  sort: "asc" | "desc";
  order_by: OrderBy;
  from_dt: string | null;
  to_dt: string | null;
  prova_id: string | null;
  usuario_id: string | null;
  acao: string | null;
  /** Filtro semantico de alto nivel (UX A2). */
  tipo_evento: TipoEvento | null;
  q: string | null;
}

export const DEFAULT_FILTERS: AuditLogFilters = {
  page: 1,
  page_size: 50,
  sort: "desc",
  order_by: "created_at",
  from_dt: null,
  to_dt: null,
  prova_id: null,
  usuario_id: null,
  acao: null,
  tipo_evento: null,
  q: null,
};

/** Converte filtros para query string ?key=value&... — pula nulos. */
export function filtersToQueryString(filters: Partial<AuditLogFilters>): string {
  const params = new URLSearchParams();
  if (filters.page !== undefined && filters.page !== null) {
    params.set("page", String(filters.page));
  }
  if (filters.page_size !== undefined && filters.page_size !== null) {
    params.set("page_size", String(filters.page_size));
  }
  if (filters.sort) {
    params.set("sort", filters.sort);
  }
  if (filters.order_by && filters.order_by !== "created_at") {
    // Omite quando default — mantem URL enxuta.
    params.set("order_by", filters.order_by);
  }
  if (filters.from_dt) {
    params.set("from", filters.from_dt);
  }
  if (filters.to_dt) {
    params.set("to", filters.to_dt);
  }
  if (filters.prova_id) {
    params.set("prova_id", filters.prova_id);
  }
  if (filters.usuario_id) {
    params.set("usuario_id", filters.usuario_id);
  }
  if (filters.acao) {
    params.set("acao", filters.acao);
  }
  if (filters.tipo_evento && filters.tipo_evento !== "todos") {
    params.set("tipo_evento", filters.tipo_evento);
  }
  if (filters.q) {
    params.set("q", filters.q);
  }
  const s = params.toString();
  return s ? `?${s}` : "";
}

// ─── Presets de data (UX A1) ─────────────────────────────────────────────

export type DatePresetKey = "hoje" | "7d" | "30d" | "90d" | "personalizado";

export const DATE_PRESET_LABELS: Record<DatePresetKey, string> = {
  hoje: "Hoje",
  "7d": "7 dias",
  "30d": "30 dias",
  "90d": "90 dias",
  personalizado: "Personalizado",
};

/** Calcula intervalo from/to em ISO UTC para um preset, baseado em now().
 *
 * "Hoje" = inicio do dia em America/Sao_Paulo ate agora.
 * "7d/30d/90d" = N dias atras (em horas corridas, do mesmo horario) ate agora.
 * "personalizado" = retorna null/null (caller mantem o que estava na URL).
 */
export function presetToRange(
  key: DatePresetKey,
  now: Date = new Date(),
): { from: string | null; to: string | null } {
  if (key === "personalizado") {
    return { from: null, to: null };
  }

  // BRT offset (-3h sem DST). Brasil aboliu DST em 2019.
  const BRT_OFFSET_MS = 3 * 60 * 60 * 1000;

  if (key === "hoje") {
    // Inicio do dia em BRT = meia-noite BRT = 03:00 UTC do mesmo dia.
    const brtNow = new Date(now.getTime() - BRT_OFFSET_MS);
    const startOfDayBrt = new Date(
      Date.UTC(
        brtNow.getUTCFullYear(),
        brtNow.getUTCMonth(),
        brtNow.getUTCDate(),
        0,
        0,
        0,
      ),
    );
    // Converte de volta para UTC (adiciona offset).
    const fromUtc = new Date(startOfDayBrt.getTime() + BRT_OFFSET_MS);
    return { from: fromUtc.toISOString(), to: now.toISOString() };
  }

  const days = key === "7d" ? 7 : key === "30d" ? 30 : 90;
  const from = new Date(now.getTime() - days * 24 * 60 * 60 * 1000);
  return { from: from.toISOString(), to: now.toISOString() };
}

/** Detecta qual preset corresponde ao intervalo atual (para destacar pill ativo).
 *
 * Tolerancia de ±1 minuto no `to` (para acomodar diferenca entre
 * "agora-quando-clicou" e "agora-quando-renderiza").
 */
export function detectPreset(
  fromDt: string | null,
  toDt: string | null,
  now: Date = new Date(),
): DatePresetKey {
  if (!fromDt || !toDt) return "personalizado";

  const tolMs = 60 * 1000;
  const toDelta = Math.abs(new Date(toDt).getTime() - now.getTime());
  if (toDelta > tolMs) return "personalizado";

  for (const key of ["hoje", "7d", "30d", "90d"] as const) {
    const expected = presetToRange(key, now);
    if (
      expected.from &&
      Math.abs(new Date(fromDt).getTime() - new Date(expected.from).getTime()) <
        tolMs
    ) {
      return key;
    }
  }
  return "personalizado";
}

// ─── Indicadores visuais ─────────────────────────────────────────────────

/** Categoriza o evento para badge colorido. Consistente com Timeline (C12). */
export type AcaoCategoria =
  | "neutro"
  | "reprovacao"
  | "reinicio"
  | "cancelamento"
  | "criacao";

/** Determina a categoria de um audit_log para coloracao na UI. */
export function categorizar(item: AuditLogItemResponse): AcaoCategoria {
  if (item.acao === "reiniciar_ciclo") {
    return "reinicio";
  }
  if (item.acao === "criar_prova") {
    return "criacao";
  }
  if (item.acao === "transitar_status" && item.detalhes_json) {
    const para = (item.detalhes_json as { para?: string }).para;
    if (para === "REPROVADA_PELO_VENDEDOR") return "reprovacao";
    if (para === "CANCELADA") return "cancelamento";
  }
  return "neutro";
}

File: backend/app/domain/schemas/audit_log.py
"""Pydantic v2 schemas para Interface de Log de Auditoria (Wave 6, Componente 18).

Implementa RNF-005: leitura imutavel do log de auditoria, restrito ao perfil
3Studio (is_admin=true). Cobre tres endpoints:
  - GET /api/v1/audit-log              -> listagem paginada com filtros
  - GET /api/v1/audit-log/{id}         -> detalhe de um registro
  - GET /api/v1/audit-log/by-prova/{id} -> historico cronologico por prova

Convencoes:
  - Datas em UTC (timezone-aware). Frontend converte para America/Sao_Paulo.
  - Campos `frozen=True` para evitar mutacao acidental no path de response.
  - `assinatura_digital` (BYTEA) NUNCA e exposta em response — apenas o
    boolean `assinatura_digital_presente` no detalhe.
  - `acao` aceita qualquer string <= 100 chars (sem hardcode dos 6 valores
    conhecidos hoje), porque migrations futuras podem adicionar novos.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.db.models import RotaEnum, SetorEnum, StatusProvaEnum


# ─── Constantes de validacao ───────────────────────────────────────────────


MAX_PAGE_SIZE = 200
"""Maximo de itens por pagina na listagem (consistente com /api/v1/users)."""

DEFAULT_PAGE_SIZE = 50
"""Default razoavel para um log auditoria (mais denso que listagem de provas)."""

MAX_RANGE_DAYS = 366
"""Maximo de dias entre `from_dt` e `to_dt` (consistente com /api/v1/reports)."""

MAX_Q_LENGTH = 200
"""Tamanho maximo da busca textual em `detalhes_json::text`."""

MAX_BY_PROVA_ITEMS = 500
"""Hard cap defensivo no historico por prova (sentinela para investigacao)."""


# ─── Query schema (params da listagem) ─────────────────────────────────────


class AuditLogListQuery(BaseModel):
    """Parametros da query GET /api/v1/audit-log.

    Todos os filtros sao opcionais. Defaults conservadores: pagina 1,
    50 itens, ordem decrescente. Validacao Pydantic gera 422 automatico.
    """

    model_config = ConfigDict(frozen=True, populate_by_name=True)

    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE)
    sort: str = Field(default="desc", pattern="^(asc|desc)$")
    """Ordenacao por created_at: 'desc' (mais recente primeiro) ou 'asc'."""

    from_dt: datetime | None = Field(default=None, alias="from")
    """Inicio do intervalo (UTC). Inclusive."""

    to_dt: datetime | None = Field(default=None, alias="to")
    """Fim do intervalo (UTC). Exclusive."""

    prova_id: UUID | None = None
    """Filtra por prova especifica."""

    usuario_id: UUID | None = None
    """Filtra por ator (usuario que executou a acao)."""

    acao: str | None = Field(default=None, max_length=100)
    """Filtra por tipo de evento. Aceita qualquer string <= 100 chars."""

    q: str | None = Field(default=None, max_length=MAX_Q_LENGTH)
    """Busca textual em detalhes_json::text (LIKE case-insensitive)."""

    @model_validator(mode="after")
    def validate_date_range(self) -> "AuditLogListQuery":
        """Valida invariantes do intervalo de datas."""
        if self.from_dt is not None and self.to_dt is not None:
            if self.from_dt >= self.to_dt:
                raise ValueError("'from' deve ser anterior a 'to'")
            if (self.to_dt - self.from_dt) > timedelta(days=MAX_RANGE_DAYS):
                raise ValueError(
                    f"Intervalo maximo permitido: {MAX_RANGE_DAYS} dias"
                )
        # Garante timezone-aware (default UTC se naive). Pydantic v2 aceita
        # naive datetimes; normalizamos para evitar comparacoes ambiguas no SQL.
        if self.from_dt is not None and self.from_dt.tzinfo is None:
            object.__setattr__(
                self, "from_dt", self.from_dt.replace(tzinfo=timezone.utc)
            )
        if self.to_dt is not None and self.to_dt.tzinfo is None:
            object.__setattr__(
                self, "to_dt", self.to_dt.replace(tzinfo=timezone.utc)
            )
        return self

    @model_validator(mode="after")
    def validate_q_no_control_chars(self) -> "AuditLogListQuery":
        """Rejeita caracteres de controle no termo de busca (defensivo)."""
        if self.q is not None:
            cleaned = self.q.strip()
            if any(ord(c) < 32 and c not in ("\t",) for c in cleaned):
                raise ValueError("Busca contem caracteres de controle invalidos")
            if cleaned == "":
                # Trata string vazia como ausencia do filtro.
                object.__setattr__(self, "q", None)
            else:
                object.__setattr__(self, "q", cleaned)
        return self


# ─── Item de listagem ──────────────────────────────────────────────────────


class AuditLogItemResponse(BaseModel):
    """Linha individual da listagem de audit_logs.

    Combina dados de audit_logs + JOINs com usuarios (nome, setor) e
    provas_digitais (nro_requerimento) para exibir labels amigaveis na UI
    sem N+1 queries.
    """

    model_config = ConfigDict(frozen=True, from_attributes=True)

    id: UUID
    acao: str
    prova_id: UUID | None
    prova_nro_requerimento: str | None
    """Nro requerimento da prova relacionada — ausente se acao nao tem prova
    (ex: atualizar_configuracao) ou se a prova foi removida (nao deve ocorrer
    em producao porque provas_digitais e auditada — mas defensivo)."""

    usuario_id: UUID
    usuario_nome: str
    usuario_setor: SetorEnum

    detalhes_json: dict[str, Any] | None
    """Payload estruturado da acao. Pode conter PII (cliente, motivo_reprovacao);
    restrito a admin via RLS pol_audit_select."""

    ip_address: str | None
    user_agent: str | None
    created_at: datetime


class AuditLogListResponse(BaseModel):
    """Response paginado da listagem.

    `total` e count(*) com os mesmos filtros aplicados (1 query extra).
    """

    model_config = ConfigDict(frozen=True)

    items: list[AuditLogItemResponse]
    total: int
    page: int
    page_size: int


# ─── Detalhe (com enriquecimento opcional de movimentacao) ────────────────


class MovimentacaoSnapshot(BaseModel):
    """Resumo da movimentacao relacionada a um audit_log de transitar_status
    ou reiniciar_ciclo.

    SEMPRE omite a `assinatura_digital` (BYTEA) — apenas expoe o boolean
    `assinatura_digital_presente` para confirmar que houve assinatura conforme
    RN-003. Caso de uso: investigador quer saber se a transicao foi assinada,
    nao quer copiar bytes.
    """

    model_config = ConfigDict(frozen=True, from_attributes=True)

    id: UUID
    status_anterior: StatusProvaEnum
    status_novo: StatusProvaEnum
    motivo_reprovacao: str | None
    ciclo: int
    rota_no_momento: RotaEnum | None
    assinatura_digital_presente: bool
    created_at: datetime


class AuditLogDetailResponse(AuditLogItemResponse):
    """Detalhe individual com enriquecimento opcional.

    `movimentacao_relacionada` e populada apenas quando `acao` e
    'transitar_status' ou 'reiniciar_ciclo' E a query de matching
    encontra exatamente uma movimentacao correspondente (ver
    audit_log_service._find_movimentacao_relacionada).
    """

    model_config = ConfigDict(frozen=True)

    movimentacao_relacionada: MovimentacaoSnapshot | None = None
